src/Sparrow.Server/Allocators.Native.cs

```diff
@@ -0,0 +1,209 @@
+                throw new NotSupportedException($"{nameof(TConfig)} is not compatible with {nameof(TOptions)}");
+
+            // This cast will get evicted by the JIT. 
+            allocator._options = (TOptions)(object)configuration;
+
+            if (((TOptions)(object)configuration).ElectricFenceEnabled && ((TOptions)(object)configuration).UseSecureMemory)
+                throw new NotSupportedException($"{nameof(TConfig)} is asking for secure, electric fenced memory. The combination is not supported.");
+        }
+
+        public long TotalAllocated
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => _totalAllocated;
+        }
+
+        public long Allocated
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => _allocated;
+        }
+
+        public long InUse
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => _allocated;
+        }
+
+        public bool IsThreadSafe => true;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Initialize(ref NativeAllocator<TOptions> allocator)
+        {
+            allocator._totalAllocated = 0;
+
+#if VALIDATE || DEBUG
+            // https://github.com/benaadams/Ben.Demystifier/issues/75
+            allocator._initializeStackTrace = EnhancedStackTrace.Current();
+            // allocator._initializeStackTrace = "currently missing because of bug";
+#endif
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Pointer Allocate(ref NativeAllocator<TOptions> allocator, int size)
+        {
+            byte* memory;
+
+            // PERF: Given that for the normal use case the INativeOptions we will use returns constants the
+            //       JIT will be able to fold all this if sequence into a branchless single call.
+            if (allocator._options.ElectricFenceEnabled)
+                memory = ElectricFencedMemory.Allocate(size);
+            else if (allocator._options.UseSecureMemory)
+                throw new NotImplementedException();
+            else
+                memory = NativeMemory.AllocateMemory(size);
+
+            if (allocator._options.Zeroed)
+                Memory.Set(memory, 0, size);
+
+            Interlocked.Add(ref allocator._allocated, size);
+            Interlocked.Add(ref allocator._totalAllocated, size);
+
+            return new Pointer(memory, size);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Release(ref NativeAllocator<TOptions> allocator, ref Pointer ptr)
+        {
+            Interlocked.Add(ref allocator._allocated, -ptr.Size);
+
+            // PERF: Given that for the normal use case the INativeOptions we will use returns constants the
+            //       JIT will be able to fold all this if sequence into a branchless single call.
```

Comment:
Do we want to zero the memory on release as well (although might be only valid for secured)?

Reply:
We can configure that from the caller site. For secure that is part of the normal operation of secure memory, so while implicitly you will setup it like false at the allocator level.